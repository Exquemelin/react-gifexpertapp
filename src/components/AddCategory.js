import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories })  => {

    // Cremoas una constante que iremos cambiando según varíen los State
    const [inputValue, setInputValue] = useState(''); // Tenemos que poner '' para que no sea un undefined y nos de un warning

    // Usamos un método para modificar la variable inputValue
    const handleInputChange = ( e ) => {
        // console.log(e.target.value);
        setInputValue(e.target.value);

        // Este Console.log es para las pruebas
        console.log('handleInputChange fue lanzado');
    }

    const handleSubmit = ( e ) => {

        // Prevenimos que al hacer submit recarge el formulario
        e.preventDefault();

        // Console.log para las pruebas
        console.log('handleSubmit lanzado', inputValue);

        // Validamos lo que nos entra por el input
        if ( inputValue.trim().length > 2 ) {

            // Activamos el setCategories que nos entra por las props
            setCategories( cats => [inputValue, ...cats] );
            // Y vaciamos el inputValue
            setInputValue('');

        }


    }
    
    // Usamos un form en lugar de un fragment, ya que también agrupa elementos y queremos hacer un formulario
    return (
        <form onSubmit={ handleSubmit }>
            <p> { inputValue } </p> 
            <input 
                type="text"
                value={ inputValue }
                onChange={ handleInputChange }
            />
        </form>
    )

}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}