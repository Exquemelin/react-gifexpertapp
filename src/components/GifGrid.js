// useEffect nos permite ejecutar cógido de forma condicional
import React from 'react' 
import { useFetchGifs } from '../hooks/useFetchGifs'
// import { getGifs } from '../helpers/getGifs';


import { GifGridItem } from './GifGridItem';



export const GifGrid = ({ category }) => {

    // const [images, setImages] = useState([]);

    // Importamos el hook y lo desestructuramos
    const { data:images, loading } = useFetchGifs( category );


    // // Condicionamos el disparo del método getGifs
    // useEffect( () => {
    //     getGifs( category )
    //     // .then( imgs => setImages( imgs ) );
    //     .then( setImages ); // al tener un solo argumento, se puede hacer más corta
    // }, [ category ]); // Si añadimos [] solo se ejecuta cuando es renderizado por primera vez
    // // Si añadimos [category] se vuelve a renderizar


    return (
        <>
            <h3 className="animate__animated animate__fadeIn"> { category } </h3>

            { loading && <p className="animate__animated animate__flash">Loading.....</p>}

            <div className="card-grid">

                {
                    images.map( gif => (
                        // <li key={ gif.id}> {gif.title} </li>
                        // console.log({ ...gif })
                        <GifGridItem 
                            key={ gif.id }
                            { ...gif }
                        />
                    ))
                }

            </div>
        </>
    )
}
