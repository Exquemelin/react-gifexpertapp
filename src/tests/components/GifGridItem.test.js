import { shallow } from "enzyme"


import { GifGridItem } from "../../components/GifGridItem"



describe('Pruebas del componente GifGridItem', () => {

    // key='Nueva Key'
    // id='nueva id'
    const url = 'nueva url';
    const title = 'nuevo title';

    // Inicializamos el wrapper como una variable para tener toda la funcionalidad
    let wrapper = shallow( <GifGridItem 
        title={ title }
        url={ url }
    /> );

    test('Debe hacer mostrar <GifGridItem />', () => {
        
        expect ( wrapper ).toMatchSnapshot();

    });

    test('Debe de tener un párrafo con el title', () => {

        const p = wrapper.find('p');
        expect( p.text().trim() ).toBe( title ) ;
        
    });

    test('debe de tener la imagen igual al url y alt de los props', () => {

        const img = wrapper.find('img');
        // console.log( img.props() ); // Usando props nos extrae las propiedades del objeto. Se podría escoger solo uno de ellos
        expect( img.prop( 'src' )).toBe( url );
        expect( img.prop( 'alt' )).toBe( title );
        
    });

    test('debe de tener animate__fadeIn', () => {

        const div = wrapper.find('div');

        // console.log(div.props());
        // console.log(div.prop('className').endsWith());

        expect( div.prop('className').includes('animate__fadeIn') ).toBe( true );

    });
    
    
    
    
    
})
