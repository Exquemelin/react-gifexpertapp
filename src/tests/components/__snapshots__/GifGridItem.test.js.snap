// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente GifGridItem Debe hacer mostrar <GifGridItem /> 1`] = `
<div
  className="animate__animated animate__fadeIn"
>
  <img
    alt="nuevo title"
    src="nueva url"
  />
  <p>
     
    nuevo title
     
  </p>
</div>
`;
